src/orthonormal_bases.jl

```diff
@@ -0,0 +1,185 @@
+
+"""
+    AbstractBasis
+
+Abstract type that represents a basis on a manifold or a subset of it.
+"""
+abstract type AbstractBasis end
+
+"""
+    AbstractOrthonormalBasis
+
+Abstract type that represents an orthonormal basis on a manifold or a subset of it.
+"""
+abstract type AbstractOrthonormalBasis <: AbstractBasis end
+
+"""
+    AbstractOrthonormalBasis
+
+Abstract type that represents an orthonormal basis of the tangent space at a point
+on a manifold. Stores tangent vectors in field `.vectors`.
+"""
+abstract type AbstractPrecomputedOrthonormalBasis <: AbstractOrthonormalBasis end
+
+"""
+    ArbitraryOrthonormalBasis
+
+An arbitrary orthonormal basis on a manifold. This will usually
+be the fastest OrthonormalBasis available for a manifold.
+"""
+struct ArbitraryOrthonormalBasis <: AbstractOrthonormalBasis end
+
+"""
+    ProjectedOrthonormalBasis(method::Symbol)
+
+An orthonormal basis that comes from orthonormalization of basis vectors
+of the ambient space projected onto the subspace representing the tangent space
+at a given point.
+
+Available methods:
+  - `:gram_schmidt` TODO
+  - `:svd` uses SVD decomposition to orthogonalize projected vectors.
+    The SVD-based method should be more numerically stable at the cost of
+    an additional assumption (local metric tensor at a point where the
+    basis is calculated has to be diagonal).
+"""
+struct ProjectedOrthonormalBasis{Method} <: AbstractOrthonormalBasis end
+
+ProjectedOrthonormalBasis(method::Symbol) = ProjectedOrthonormalBasis{method}()
+
+"""
+    PrecomputedOrthonormalBasis(vectors::AbstractVector)
+
+A precomputed orthonormal basis at a point on a manifold.
+"""
+struct PrecomputedOrthonormalBasis{TV<:AbstractVector} <: AbstractPrecomputedOrthonormalBasis
+    vectors::TV
+end
+
+"""
+    DiagonalizingOrthonormalBasis(vectors, kappas)
+"""
+struct DiagonalizingOrthonormalBasis{TV<:AbstractVector, TK<:AbstractVector} <: AbstractPrecomputedOrthonormalBasis
+    vectors::TV
+    kappas::TK
+end
+
+"""
+    represent_in_basis(M::Manifold, x, v, B::AbstractBasis)
+
+Compute a one-dimentional vector of coefficients of the tangent vector `v`
+at point denoted by `x` on manifold `M` in basis `B`.
+
+Depending on the basis, `x` may not directly represent a point on the manifold.
+For example if a basis transported along a curve is used, `x` may be the coordinate
+along the curve.
+
+See also: [`inverse_represent_in_basis`](@ref), [`basis`](@ref)
+"""
+function represent_in_basis(M::Manifold, x, v, B::AbstractBasis)
+    error("represent_in_basis not implemented for manifold of type $(typeof(M)) a point of type $(typeof(x)), tangent vector of type $(typeof(v)) and basis of type $(typeof(B)).")
+end
+
+function represent_in_basis(M::Manifold, x, v, B::PrecomputedOrthonormalBasis)
+    return map(vb -> inner(M, x, v, vb), B.vectors)
```

Comment:
Here and below, this will need a call to `real` to get real coefficients on manifolds not on a real field.

Reply:
I've fixed this one but where is the other real needed below? I don't see it.